pkg/reconciler/events/cloudevent/cloud_event_controller.go

```diff
@@ -140,6 +140,7 @@ func SendCloudEventWithRetries(ctx context.Context, object runtime.Object) error
 	wasIn := make(chan error)
 	go func() {
 		wasIn <- nil
+		logger.Infof("Sending cloudevent of type %q", event.Type())
```

Comment:
If this is meant as a permanent addition, I would suggest to log it with `debug` level; this may add almost one line of log per reconcile, which in a busy cluster is a lot.

Reply:
good call! just double checking, will debugf show up in test log output when a test fails?

Reply:
Thanks a very good question.
I downloaded this PR and tried inducing a forced failure:
```
$ git diff
diff --git a/pkg/reconciler/taskrun/taskrun_test.go b/pkg/reconciler/taskrun/taskrun_test.go
index b31e8658..9b2fa939 100644
--- a/pkg/reconciler/taskrun/taskrun_test.go
+++ b/pkg/reconciler/taskrun/taskrun_test.go
@@ -790,6 +790,7 @@ func TestReconcile_CloudEvents(t *testing.T) {
        if !(err == nil) {
                t.Errorf(err.Error())
        }
+       t.Errorf("Forced failure")
 }

 func TestReconcile(t *testing.T) {
```

The log from the failed test does include `DEBUG` statements, but not the one we're looking for. It also looks like the last logging statement is using a different logger, so I wonder if something is not ok in the logger setup either in the cloud events package or in the unit tests:
```
$ go test -race -v ./... -run '^TestReconcile_CloudEvents'
=== RUN   TestReconcile_CloudEvents
2020-09-25T08:18:15.660+0100	INFO	TestReconcile_CloudEvents	taskrun/controller.go:93	Setting up event handlers
2020-09-25T08:18:15.662+0100	INFO	TestReconcile_CloudEvents.config-store	configmap/store.go:154	defaults/features/artifacts config "config-defaults" config was added or updated: &config.Defaults{DefaultTimeoutMinutes:60, DefaultServiceAccount:"default", DefaultManagedByLabelValue:"tekton-pipelines", DefaultPodTemplate:(*pod.Template)(nil), DefaultCloudEventsSink:"http://synk:8080", DefaultTaskRunWorkspaceBinding:""}
2020-09-25T08:18:15.663+0100	INFO	TestReconcile_CloudEvents.config-store	configmap/store.go:154	defaults/features/artifacts config "feature-flags" config was added or updated: &config.FeatureFlags{DisableHomeEnvOverwrite:false, DisableWorkingDirOverwrite:false, DisableAffinityAssistant:false, RunningInEnvWithInjectedSidecars:true}
2020-09-25T08:18:15.663+0100	INFO	TestReconcile_CloudEvents.config-store	configmap/store.go:154	defaults/features/artifacts config "config-artifact-bucket" config was added or updated: &config.ArtifactBucket{Location:"", ServiceAccountSecretName:"", ServiceAccountSecretKey:"", ServiceAccountFieldName:"GOOGLE_APPLICATION_CREDENTIALS"}
2020-09-25T08:18:15.663+0100	INFO	TestReconcile_CloudEvents.config-store	configmap/store.go:154	defaults/features/artifacts config "config-artifact-pvc" config was added or updated: &config.ArtifactPVC{Size:"5Gi", StorageClassName:""}
{"level":"info","ts":1601018295.775664,"logger":"fallback","caller":"taskrun/taskrun.go:320","msg":"Cloud Events: []"}
2020-09-25T08:18:15.778+0100	INFO	TestReconcile_CloudEvents	timeout/handler.go:243	About to start timeout timer for foo/test-taskrun-not-started. started at 2020-09-25 08:18:15.765377 +0100 BST m=+0.177450986, timeout is {1h0m0s}, running for 12.764516ms
2020-09-25T08:18:15.778+0100	DEBUG	TestReconcile_CloudEvents.github.com-tektoncd-pipeline-pkg-reconciler-taskrun.Reconciler	controller/controller.go:394	Adding to queue foo/test-taskrun-not-started (depth: 1)	{"knative.dev/key": {"knative.dev/key": "foo/test-taskrun-not-started"}}
{"level":"info","ts":1601018295.778663,"logger":"fallback","caller":"taskrun/taskrun.go:410","msg":"Successfully reconciled taskrun test-taskrun-not-started/foo with status: &apis.Condition{Type:\"Succeeded\", Status:\"Unknown\", Severity:\"\", LastTransitionTime:apis.VolatileTime{Inner:v1.Time{Time:time.Time{wall:0xbfd3844dee682080, ext:190649436, loc:(*time.Location)(0x50e49c0)}}}, Reason:\"Running\", Message:\"Not all Steps in the Task have finished executing\"}"}
    TestReconcile_CloudEvents: taskrun_test.go:793: Forced failure
--- FAIL: TestReconcile_CloudEvents (2.14s)
2020-09-25T08:18:17.799+0100	INFO	TestReconcile_CloudEvents	timeout/handler.go:268	stopping timer for "foo/test-taskrun-not-started"
FAIL
FAIL	github.com/tektoncd/pipeline/pkg/reconciler/taskrun	3.123s
testing: warning: no tests to run
PASS
ok  	github.com/tektoncd/pipeline/pkg/reconciler/taskrun/resources	0.836s [no tests to run]
FAIL
```

Reply:
So, it looks like the problem is that we setup a context in the test assets, but do not return it, so when we invoke `Reconcile` the only option we have is to setup a fresh new `context.Background` which uses the fallback logger.

I tried the following changes:
```
$ git diff
diff --git a/pkg/reconciler/taskrun/taskrun_test.go b/pkg/reconciler/taskrun/taskrun_test.go
index b31e8658..48d588de 100644
--- a/pkg/reconciler/taskrun/taskrun_test.go
+++ b/pkg/reconciler/taskrun/taskrun_test.go
@@ -300,6 +300,7 @@ func getTaskRunController(t *testing.T, d test.Data) (test.Assets, func()) {
                Clients:    c,
                Informers:  informers,
                Recorder:   controller.GetEventRecorder(ctx).(*record.FakeRecorder),
+               Ctx:        ctx,
        }, cancel
 }

@@ -753,7 +754,7 @@ func TestReconcile_CloudEvents(t *testing.T) {
                t.Fatal(err)
        }

-       if err := c.Reconciler.Reconcile(context.Background(), getRunName(taskRun)); err != nil {
+       if err := c.Reconciler.Reconcile(testAssets.Ctx, getRunName(taskRun)); err != nil {
                t.Errorf("expected no error. Got error %v", err)
        }
        if len(clients.Kube.Actions()) == 0 {
@@ -790,6 +791,7 @@ func TestReconcile_CloudEvents(t *testing.T) {
        if !(err == nil) {
                t.Errorf(err.Error())
        }
+       t.Errorf("Forced failure")
 }

 func TestReconcile(t *testing.T) {
diff --git a/test/controller.go b/test/controller.go
index 939b79b6..2c4fdf12 100644
--- a/test/controller.go
+++ b/test/controller.go
@@ -102,6 +102,7 @@ type Assets struct {
        Clients    Clients
        Informers  Informers
        Recorder   *record.FakeRecorder
+       Ctx        context.Context
 }
```

and now the log line shows up:
```
$ go test -race -v ./... -run '^TestReconcile_CloudEvents'
=== RUN   TestReconcile_CloudEvents
2020-09-25T08:41:29.629+0100	INFO	TestReconcile_CloudEvents	taskrun/controller.go:93	Setting up event handlers
2020-09-25T08:41:29.631+0100	INFO	TestReconcile_CloudEvents.config-store	configmap/store.go:154	defaults/features/artifacts config "config-defaults" config was added or updated: &config.Defaults{DefaultTimeoutMinutes:60, DefaultServiceAccount:"default", DefaultManagedByLabelValue:"tekton-pipelines", DefaultPodTemplate:(*pod.Template)(nil), DefaultCloudEventsSink:"http://synk:8080", DefaultTaskRunWorkspaceBinding:""}
2020-09-25T08:41:29.632+0100	INFO	TestReconcile_CloudEvents.config-store	configmap/store.go:154	defaults/features/artifacts config "feature-flags" config was added or updated: &config.FeatureFlags{DisableHomeEnvOverwrite:false, DisableWorkingDirOverwrite:false, DisableAffinityAssistant:false, RunningInEnvWithInjectedSidecars:true}
2020-09-25T08:41:29.632+0100	INFO	TestReconcile_CloudEvents.config-store	configmap/store.go:154	defaults/features/artifacts config "config-artifact-bucket" config was added or updated: &config.ArtifactBucket{Location:"", ServiceAccountSecretName:"", ServiceAccountSecretKey:"", ServiceAccountFieldName:"GOOGLE_APPLICATION_CREDENTIALS"}
2020-09-25T08:41:29.632+0100	INFO	TestReconcile_CloudEvents.config-store	configmap/store.go:154	defaults/features/artifacts config "config-artifact-pvc" config was added or updated: &config.ArtifactPVC{Size:"5Gi", StorageClassName:""}
2020-09-25T08:41:29.746+0100	DEBUG	TestReconcile_CloudEvents	cloudevent/cloud_event_controller.go:143	Sending cloudevent of type "dev.tekton.event.taskrun.started.v1"
2020-09-25T08:41:29.746+0100	INFO	TestReconcile_CloudEvents	taskrun/taskrun.go:320	Cloud Events: []
2020-09-25T08:41:29.748+0100	INFO	TestReconcile_CloudEvents	timeout/handler.go:243	About to start timeout timer for foo/test-taskrun-not-started. started at 2020-09-25 08:41:29.734605 +0100 BST m=+0.177344374, timeout is {1h0m0s}, running for 14.254145ms
2020-09-25T08:41:29.748+0100	DEBUG	TestReconcile_CloudEvents.github.com-tektoncd-pipeline-pkg-reconciler-taskrun.Reconciler	controller/controller.go:394	Adding to queue foo/test-taskrun-not-started (depth: 1)	{"knative.dev/key": {"knative.dev/key": "foo/test-taskrun-not-started"}}
2020-09-25T08:41:29.749+0100	INFO	TestReconcile_CloudEvents	taskrun/taskrun.go:410	Successfully reconciled taskrun test-taskrun-not-started/foo with status: &apis.Condition{Type:"Succeeded", Status:"Unknown", Severity:"", LastTransitionTime:apis.VolatileTime{Inner:v1.Time{Time:time.Time{wall:0xbfd385aa6ca5cb58, ext:191802149, loc:(*time.Location)(0x50e49c0)}}}, Reason:"Running", Message:"Not all Steps in the Task have finished executing"}
2020-09-25T08:41:29.749+0100	DEBUG	TestReconcile_CloudEvents	cloudevent/cloud_event_controller.go:143	Sending cloudevent of type "dev.tekton.event.taskrun.running.v1"
2020-09-25T08:41:29.750+0100	DEBUG	TestReconcile_CloudEvents	taskrun/reconciler.go:322	Updating status with:   v1beta1.TaskRunStatus{
  	Status: v1beta1.Status{
  		ObservedGeneration: 0,
- 		Conditions:         nil,
+ 		Conditions: v1beta1.Conditions{
+ 			{
+ 				Type:               "Succeeded",
+ 				Status:             "Unknown",
+ 				LastTransitionTime: apis.VolatileTime{Inner: s"2020-09-25 08:41:29.749063 +0100"...},
+ 				Reason:             "Running",
+ 				Message:            "Not all Steps in the Task have finished executing",
+ 			},
+ 		},
  		Annotations: nil,
  	},
  	TaskRunStatusFields: v1beta1.TaskRunStatusFields{
- 		PodName:         "",
+ 		PodName:         "test-taskrun-not-started-pod-mz4c7",
- 		StartTime:       nil,
+ 		StartTime:       s"2020-09-25 08:41:29.734605 +0100 BST m=+0.177344374",
  		CompletionTime:  nil,
- 		Steps:           nil,
+ 		Steps:           []v1beta1.StepState{},
  		CloudEvents:     nil,
  		RetriesStatus:   nil,
  		ResourcesResult: nil,
  		TaskRunResults:  nil,
- 		Sidecars:        nil,
+ 		Sidecars:        []v1beta1.SidecarState{},
- 		TaskSpec:        nil,
+ 		TaskSpec:        &v1beta1.TaskSpec{Steps: []v1beta1.Step{{Container: v1.Container{...}}}},
  	},
  }

2020-09-25T08:41:29.763+0100	DEBUG	TestReconcile_CloudEvents	taskrun/reconciler.go:322	Updating status with:   v1beta1.TaskRunStatus{
  	Status: v1beta1.Status{
  		ObservedGeneration: 0,
- 		Conditions:         nil,
+ 		Conditions: v1beta1.Conditions{
+ 			{
+ 				Type:               "Succeeded",
+ 				Status:             "Unknown",
+ 				LastTransitionTime: apis.VolatileTime{Inner: s"2020-09-25 08:41:29.749063 +0100"...},
+ 				Reason:             "Running",
+ 				Message:            "Not all Steps in the Task have finished executing",
+ 			},
+ 		},
  		Annotations: nil,
  	},
  	TaskRunStatusFields: v1beta1.TaskRunStatusFields{
- 		PodName:         "",
+ 		PodName:         "test-taskrun-not-started-pod-mz4c7",
- 		StartTime:       nil,
+ 		StartTime:       s"2020-09-25 08:41:29.734605 +0100 BST m=+0.177344374",
  		CompletionTime:  nil,
- 		Steps:           nil,
+ 		Steps:           []v1beta1.StepState{},
  		CloudEvents:     nil,
  		RetriesStatus:   nil,
  		ResourcesResult: nil,
  		TaskRunResults:  nil,
- 		Sidecars:        nil,
+ 		Sidecars:        []v1beta1.SidecarState{},
- 		TaskSpec:        nil,
+ 		TaskSpec:        &v1beta1.TaskSpec{Steps: []v1beta1.Step{{Container: v1.Container{...}}}},
  	},
  }

    TestReconcile_CloudEvents: taskrun_test.go:794: Forced failure
--- FAIL: TestReconcile_CloudEvents (2.14s)
2020-09-25T08:41:31.768+0100	INFO	TestReconcile_CloudEvents	timeout/handler.go:268	stopping timer for "foo/test-taskrun-not-started"
FAIL
FAIL	github.com/tektoncd/pipeline/pkg/reconciler/taskrun	3.702s
testing: warning: no tests to run
PASS
ok  	github.com/tektoncd/pipeline/pkg/reconciler/taskrun/resources	(cached) [no tests to run]
FAIL
````

Reply:
I think we should merge this PR as it is, and have a separate one which changes *all* of our reconcile tests to call `Reconcile` with the correct test context.

Reply:
...and it is merged already :D